pennylane/templates/subroutines/out_poly.py

```diff
@@ -0,0 +1,391 @@
+        In the usage details it is shown an example where a specific modulus is used.
+
+
+    .. seealso:: :class:`~.PhaseAdder`
+
+    .. details::
+        :title: Usage Details
+
+        This template can take a modulus different from powers of two. In these cases it should be provided two auxiliary qubits.
+
+        .. code-block:: python
+
+            x_wires = [0, 1, 2]
+            y_wires = [3, 4, 5]
+            output_wires = [6, 7, 8]
+            work_wires = [9,10]
+
+
+            def f(x, y):
+                return x ** 2 + y
+
+            @qml.qnode(qml.device("default.qubit", shots = 1))
+            def circuit():
+                # loading values for x and y
+                qml.BasisEmbedding(3, wires=wires_x)
+                qml.BasisEmbedding(2, wires=wires_y)
+
+                # applying the polynomial
+                qml.OutPoly(
+                    f,
+                    x_wires = x_wires,
+                    y_wires = y_wires,
+                    output_wires = output_wires,
+                    mod = 7,
+                    work_wires = work_wires)
+
+                return qml.sample(wires=output_wires)
+
+        .. code-block:: pycon
+
+            >>> print(circuit())
+            [1 0 0]
+
+        The result, :math:`[1 0 0]`, is the binary representation of :math:`3^2 + 2  \; \text{modulo} \; 7 = 4`.
+        If the output wires are not initialized to zero, this value will be added to the solution. Generically, the expression is definded as:
+
+        .. math::
+
+            \text{OutPoly}_{f, \text{mod}} |x_1 \rangle \dots |x_m \rangle |b \rangle
+            = |x_1 \rangle \dots |x_m \rangle |b + f(x_1, \dots, x_m) \mod \text{mod} \rangle.
+
+    """
+
+    grad_method = None
+
+    def __init__(
+        self, f, output_wires, mod=None, work_wires=None, id=None, **kwargs
+    ):  # pylint: disable=too-many-arguments
+        r"""Initialize the OutPoly class"""
+
+        registers_wires = [*kwargs.values(), output_wires]
+
+        if not registers_wires or f is None:
+            raise ValueError("The register wires and the function f must be provided.")
+
+        num_work_wires = 0 if not work_wires else len(work_wires)
+        if mod is None:
+            mod = 2 ** len(registers_wires[-1])
+        elif mod != 2 ** len(registers_wires[-1]) and num_work_wires != 2:
+            raise ValueError(
+                f"If mod is not 2^{len(registers_wires[-1])}, two work wires should be provided"
+            )
+
+        if not isinstance(mod, int):
+            raise ValueError("mod must be integer.")
+
+        self.hyperparameters["f"] = f
+
+        # if len(inspect.signature(f).parameters) != len(registers_wires) - 1:
+        #    raise ValueError(
+        #        f"The function takes {len(inspect.signature(f).parameters)} input parameters but {len(registers_wires) - 1} has provided."
+        #    )
```

Comment:
For the moment I leave it commented, pending to confirm with @isaacdevlugt  if we want me to check this or not

Not including it may give flexibility on a large scale but may help the user to detect errors